File: packages/desktop/src/lib/acp/types/interaction.ts
import type { PermissionRequest } from "./permission.js";
import type { AnsweredQuestion, QuestionRequest } from "./question.js";
import type { InteractionReplyHandler } from "./reply-handler.js";
import type { ComputerPermissionKind } from "../../services/acp-types.js";

export type InteractionKind = "permission" | "question" | "plan_approval" | "computer_permission";

export interface InteractionToolReference {
	messageID: string | null;
	callID: string;
}

export interface PermissionInteraction {
	id: string;
	kind: "permission";
	request: PermissionRequest;
}

export interface QuestionInteraction {
	id: string;
	kind: "question";
	request: QuestionRequest;
	answered?: AnsweredQuestion;
}

export interface PlanApprovalInteraction {
	id: string;
	kind: "plan_approval";
	source: "create_plan" | "exit_plan_mode";
	sessionId: string;
	tool: InteractionToolReference;
	jsonRpcRequestId?: number;
	replyHandler: InteractionReplyHandler;
	status: "pending" | "approved" | "rejected";
	canonicalOperationId?: string | null;
}

export interface ComputerPermissionInteraction {
	id: string;
	kind: "computer_permission";
	sessionId: string;
	permissionKind: ComputerPermissionKind;
	reason: string;
	app?: string | null;
	window?: string | null;
	tool?: InteractionToolReference;
	status: "pending" | "approved" | "rejected";
	canonicalOperationId?: string | null;
}

export type Interaction =
	| PermissionInteraction
	| QuestionInteraction
	| PlanApprovalInteraction
	| ComputerPermissionInteraction;

export function buildPlanApprovalInteractionId(
	sessionId: string,
	toolCallId: string,
	jsonRpcRequestId: number
): string {
	return `${sessionId}\u0000${toolCallId}\u0000plan\u0000${jsonRpcRequestId}`;
}

File: packages/desktop/src-tauri/src/acp/projections/types/interaction.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use specta::Type;

use crate::acp::session_update::{
    InteractionReplyHandler, PermissionData, QuestionData, ToolReference,
};
use crate::computer_use::permissions::ComputerPermissionKind;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub enum InteractionKind {
    Permission,
    Question,
    PlanApproval,
    ComputerPermission,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub enum InteractionState {
    Pending,
    Approved,
    Rejected,
    Answered,
    Unresolved,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub enum PlanApprovalSource {
    CreatePlan,
    ExitPlanMode,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct ComputerPermissionData {
    pub id: String,
    pub session_id: String,
    pub permission_kind: ComputerPermissionKind,
    pub reason: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub app: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub window: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub tool: Option<ToolReference>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub enum InteractionPayload {
    Permission(PermissionData),
    Question(QuestionData),
    PlanApproval { source: PlanApprovalSource },
    ComputerPermission(ComputerPermissionData),
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "snake_case", tag = "kind")]
pub enum InteractionResponse {
    Permission {
        accepted: bool,
        #[serde(skip_serializing_if = "Option::is_none")]
        option_id: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        reply: Option<String>,
    },
    Question {
        answers: Value,
    },
    PlanApproval {
        approved: bool,
    },
    ComputerPermission {
        accepted: bool,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct InteractionSnapshot {
    pub id: String,
    pub session_id: String,
    pub kind: InteractionKind,
    pub state: InteractionState,
    pub json_rpc_request_id: Option<u64>,
    pub reply_handler: Option<InteractionReplyHandler>,
    pub tool_reference: Option<ToolReference>,
    pub responded_at_event_seq: Option<i64>,
    pub response: Option<InteractionResponse>,
    pub payload: InteractionPayload,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub canonical_operation_id: Option<String>,
}
